numpy/core/src/multiarray/ctors.c

```diff
@@ -477,32 +472,40 @@ setArrayFromSequence(PyArrayObject *a, PyObject *s,
         return 0;
     }
 
+    /*
+     * This code is to ensure that the sequence access below will
+     * return a lower-dimensional sequence.
+     */
+
     if (dim > PyArray_NDIM(a)) {
         PyErr_Format(PyExc_ValueError,
                  "setArrayFromSequence: sequence/array dimensions mismatch.");
         goto fail;
     }
 
-    tmp = _array_from_array_like(s, /*dtype*/NULL, /*writeable*/0, /*context*/NULL);
-    if (tmp == NULL) {
+    slen = PySequence_Length(s);
+    if (slen < 0) {
         goto fail;
     }
-    else if (tmp != Py_NotImplemented) {
-        if (PyArray_CopyInto(dst, (PyArrayObject *)tmp) < 0) {
+    if (slen > 0) {
+        /* gh-13659: objects with 0 len can still have ndim > ndim(dst) */
+        tmp = _array_from_array_like(s, /*dtype*/NULL, /*writeable*/0, /*context*/NULL);
+        if (tmp == NULL) {
             goto fail;
         }
+        else if (tmp != Py_NotImplemented) {
+            if (PyArray_CopyInto(dst, (PyArrayObject *)tmp) < 0) {
+                goto fail;
+            }
 
-        Py_DECREF(s);
-        return 0;
-    }
-    else {
-        Py_DECREF(Py_NotImplemented);
+            Py_DECREF(s);
+            return 0;
+        }
+        else {
+            Py_DECREF(Py_NotImplemented);
```

Comment:
Shouldn't this be an unconditional `Py_DECREF(tmp)`?

Reply:
fixing